File: 605-can-place-flowers/can-place-flowers.cpp
class Solution {
public:
    bool canPlaceFlowers(vector<int>& flowerbed, int n) {
        if (n == 0) {
            return true;
        }

        int i = 0;
        int length = flowerbed.size();

        while (i < length) {
            if (flowerbed[i] == 1) {
                i += 2;
                continue;
            }

            if ((i == 0 || flowerbed[i - 1] == 0) && (i == length - 1 || flowerbed[i + 1] == 0)) {
                flowerbed[i] = 1;
                n--;

                if (n == 0) {
                    return true;
                }

                i += 2;
            } else {
                i++;
            }
        }
        return false;
    }
};

File: 605-can-place-flowers/can-place-flowers.rs
impl Solution {
    pub fn can_place_flowers(flowerbed: Vec<i32>, n: i32) -> bool {
        let mut k = n;
        let mut temp = flowerbed;
        if k == 0 {
            return true;
        }

        let length = temp.len();
        let mut i = 0;

        while i < length {
            if temp[i] == 1 {
                i += 2;
                continue;
            }

            let prev_empty = i == 0 || temp[i - 1] == 0;
            let next_empty = i == length - 1 || temp[i + 1] == 0;

            if prev_empty && next_empty {
                temp[i] = 1;
                k -= 1;

                if k == 0 {
                    return true;
                }

                i += 2;
            } else {
                i += 1;
            }
        }

        return false;
    }
}